File: Platforms/Windows/Reign_Video_D3D9_Component/VertexBuffer.cpp
#include "pch.h"
#include "VertexBuffer.h"

namespace Reign_Video_D3D9_Component
{
	#pragma region Constructors
	VertexBufferCom::VertexBufferCom(VideoCom^ video, REIGN_D3DPRIMITIVETYPE topology)
	{
		this->video = video;
		SetTopology(topology);
	}

	VertexBufferCom::~VertexBufferCom()
	{
		if (vertexBuffer) vertexBuffer->Release();
		null();
	}

	void VertexBufferCom::null()
	{
		vertexBuffer = 0;
	}
	#pragma endregion

	#pragma region Methods
	void VertexBufferCom::SetTopology(REIGN_D3DPRIMITIVETYPE topology)
	{
		this->topology = (D3DPRIMITIVETYPE)topology;
		switch (topology)
		{
			case (REIGN_D3DPRIMITIVETYPE::TRIANGLELIST): primitiveVertexCount = 3; break;
			case (REIGN_D3DPRIMITIVETYPE::LINELIST): primitiveVertexCount = 2; break;
			case (REIGN_D3DPRIMITIVETYPE::POINTLIST): primitiveVertexCount = 1; break;
		}
	}

	VertexBufferErrors VertexBufferCom::Init(array<float>^ vertices, REIGN_D3DUSAGE usage, int vertexCount, int vertexByteSize)
	{
		if (vertexBuffer)
		{
			vertexBuffer->Release();
			vertexBuffer = 0;
		}

		uint bufferSize = vertexCount * vertexByteSize;
		IDirect3DVertexBuffer9* vertexBufferTEMP;
		if (FAILED(video->device->CreateVertexBuffer(bufferSize, D3DUSAGE_WRITEONLY, 0, video->caps->ExDevice ? D3DPOOL_DEFAULT : D3DPOOL_MANAGED, &vertexBufferTEMP, 0)))
		{
			return VertexBufferErrors::VertexBuffer;
		}
		vertexBuffer = vertexBufferTEMP;
		Update(vertices, vertexCount, vertexByteSize);

		return VertexBufferErrors::None;
	}

	void VertexBufferCom::Update(array<float>^ vertices, int updateCount, int vertexByteSize)
	{
		void *data = 0;
		uint bufferSize = vertexByteSize * updateCount;
		vertexBuffer->Lock(0, bufferSize, (void**)&data, 0);
		pin_ptr<float>verticesTEMP = &vertices[0];
		memcpy(data, verticesTEMP, bufferSize);
		vertexBuffer->Unlock();
	}

	void VertexBufferCom::Enable(IndexBufferCom^ indexBuffer, VertexBufferCom^ instanceBuffer, int vertexByteSize, int instanceVertexByteSize)
	{
		video->device->SetStreamSource(0, vertexBuffer, 0, vertexByteSize);

		if (instanceBuffer)
		{
			video->device->SetStreamSourceFreq(1, D3DSTREAMSOURCE_INSTANCEDATA | 1);
			video->device->SetStreamSource(1, instanceBuffer->vertexBuffer, 0, instanceVertexByteSize);
		}
		else
		{
			video->device->SetStreamSourceFreq(0, 1);
			video->device->SetStreamSourceFreq(1, 1);
		}

		if (indexBuffer)
		{
			this->indexBuffer = indexBuffer;
			this->indexBuffer->enable();
		}
		else
		{
			this->indexBuffer = nullptr;
		}
	}

	void VertexBufferCom::Draw(int drawCount, int vertexCount, int indexCount)
	{
		if (indexBuffer == nullptr) video->device->DrawPrimitive(topology, 0, (vertexCount/primitiveVertexCount));
		else video->device->DrawIndexedPrimitive(topology, 0, 0, vertexCount, 0, (indexCount/primitiveVertexCount));
	}

	void VertexBufferCom::Draw(int drawCount, int vertexCount)
	{
		if (indexBuffer == nullptr) video->device->DrawPrimitive(topology, 0, (drawCount/primitiveVertexCount));
		else video->device->DrawIndexedPrimitive(topology, 0, 0, vertexCount, 0, (drawCount/primitiveVertexCount));
	}

	void VertexBufferCom::DrawInstanced(int drawCount, int vertexCount)
	{
		video->device->SetStreamSourceFreq(0, D3DSTREAMSOURCE_INDEXEDDATA | drawCount);
		Draw(drawCount, vertexCount);
	}

	void VertexBufferCom::DrawInstancedClassic(int drawCount, int meshVertexCount, int meshIndexCount)
	{
		if (indexBuffer == nullptr) video->device->DrawPrimitive(topology, 0, drawCount * (meshVertexCount/primitiveVertexCount));
		else video->device->DrawIndexedPrimitive(topology, 0, 0, drawCount * meshVertexCount, 0, drawCount * (meshIndexCount/primitiveVertexCount));
	}
	#pragma endregion
}

File: Platforms/Win32/Reign.Video.D3D9/Buffers/VertexBuffer.cs
using System;
using Reign_Video_D3D9_Component;
using Reign.Video;
using Reign.Core;

namespace Reign.Video.D3D9
{
	public class VertexBuffer : IVertexBuffer
	{
		#region Properties
		private VertexBufferCom com;

		private VertexBufferTopologys topology;
		public override VertexBufferTopologys Topology
		{
			get {return topology;}
			set
			{
				switch (value)
				{
					case VertexBufferTopologys.Triangle: com.SetTopology(REIGN_D3DPRIMITIVETYPE.TRIANGLELIST); break;
					case VertexBufferTopologys.Line: com.SetTopology(REIGN_D3DPRIMITIVETYPE.LINELIST); break;
					case VertexBufferTopologys.Point: com.SetTopology(REIGN_D3DPRIMITIVETYPE.POINTLIST); break;
				}
				topology = value;
			}
		}

		private IndexBuffer indexBuffer, currentIndexBuffer;
		public override IIndexBuffer IndexBuffer
		{
			get {return indexBuffer;}
		}
		#endregion

		#region Constructors
		public VertexBuffer(IDisposableResource parent, IBufferLayoutDesc bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices)
		: base(parent, bufferLayoutDesc, usage, vertices)
		{
			init(parent, bufferLayoutDesc, usage, topology, vertices, null);
		}

		public VertexBuffer(IDisposableResource parent, IBufferLayoutDesc bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices, int[] indices)
		: base(parent, bufferLayoutDesc, usage, vertices)
		{
			init(parent, bufferLayoutDesc, usage, topology, vertices, indices);
		}

		private void init(IDisposableResource parent, IBufferLayoutDesc bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices, int[] indices)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				this.topology = topology;
				REIGN_D3DPRIMITIVETYPE topologyType = REIGN_D3DPRIMITIVETYPE.TRIANGLELIST;
				switch (topology)
				{
					case VertexBufferTopologys.Triangle: topologyType = REIGN_D3DPRIMITIVETYPE.TRIANGLELIST; break;
					case VertexBufferTopologys.Line: topologyType = REIGN_D3DPRIMITIVETYPE.LINELIST; break;
					case VertexBufferTopologys.Point: topologyType = REIGN_D3DPRIMITIVETYPE.POINTLIST; break;
				}

				com = new VertexBufferCom(video.com, topologyType);
				var error = com.Init(vertices, REIGN_D3DUSAGE.WRITEONLY, vertexCount, vertexByteSize);
				if (error == VertexBufferErrors.VertexBuffer) Debug.ThrowError("VertexBuffer", "Failed to create VertexBuffer");
				
				if (indices != null && indices.Length != 0) indexBuffer = new IndexBuffer(this, usage, indices);
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public override void Update(float[] vertices, int updateCount)
		{
			com.Update(vertices, updateCount, vertexByteSize);
		}

		public override void Enable()
		{
			currentIndexBuffer = indexBuffer;
			if (indexBuffer != null) com.Enable(indexBuffer.com, null, vertexByteSize, 0);
			else com.Enable(null, null, vertexByteSize, 0);
		}

		public override void Enable(IIndexBuffer indexBuffer)
		{
			this.currentIndexBuffer = (IndexBuffer)indexBuffer;
			com.Enable(((IndexBuffer)indexBuffer).com, null, vertexByteSize, 0);
		}

		public override void Enable(IVertexBuffer instanceBuffer)
		{
			currentIndexBuffer = indexBuffer;
			var ib = ((VertexBuffer)instanceBuffer);
			if (indexBuffer != null) com.Enable(indexBuffer.com, ib.com, vertexByteSize, ib.vertexByteSize);
			else com.Enable(null, ib.com, vertexByteSize, ib.vertexByteSize);
		}

		public override void Enable(IIndexBuffer indexBuffer, IVertexBuffer instanceBuffer)
		{
			this.currentIndexBuffer = (IndexBuffer)indexBuffer;
			var ib = ((VertexBuffer)instanceBuffer);
			com.Enable(((IndexBuffer)indexBuffer).com, ib.com, vertexByteSize, ib.vertexByteSize);
		}

		public override void Draw()
		{
			com.Draw(vertexCount, vertexCount, currentIndexBuffer.IndexCount);
		}

		public override void Draw(int drawCount)
		{
			com.Draw(drawCount, vertexCount);
		}

		public override void DrawInstanced(int drawCount)
		{
			com.DrawInstanced(drawCount, vertexCount);
		}

		public override void DrawInstancedClassic(int drawCount, int meshVertexCount, int meshIndexCount)
		{
			com.DrawInstancedClassic(drawCount, meshVertexCount, meshIndexCount);
		}
		#endregion
	}
}
